core/src/main/java/io/grpc/Metadata.java

```diff
@@ -650,112 +792,4 @@ T parseBytes(byte[] serialized) {
-    Key key;
-    boolean isBinary;
-    byte[] serializedBinary;
-
-    /**
-     * Constructor used when application layer adds a parsed value.
-     */
-    private MetadataEntry(Key<?> key, Object parsed) {
-      this.parsed = Preconditions.checkNotNull(parsed, "parsed");
-      this.key = Preconditions.checkNotNull(key, "key");
-      this.isBinary = key instanceof BinaryKey;
-    }
-
-    /**
-     * Constructor used when reading a value from the transport.
-     */
-    private MetadataEntry(boolean isBinary, byte[] serialized) {
-      Preconditions.checkNotNull(serialized, "serialized");
-      this.serializedBinary = serialized;
-      this.isBinary = isBinary;
-    }
-
-    /**
-     * Copy constructor.
-     */
-    private MetadataEntry(MetadataEntry entry) {
-      this.parsed = entry.parsed;
-      this.key = entry.key;
-      this.isBinary = entry.isBinary;
-      this.serializedBinary = entry.serializedBinary;
-    }
-
-    @SuppressWarnings("unchecked")
-    public <T> T getParsed(Key<T> key) {
-      T value = (T) parsed;
-      if (value != null) {
-        if (this.key != key) {
-          // Keys don't match so serialize using the old key
-          serializedBinary = this.key.toBytes(value);
-        } else {
-          return value;
-        }
-      }
-      this.key = key;
-      if (serializedBinary != null) {
-        value = key.parseBytes(serializedBinary);
-      }
-      parsed = value;
-      return value;
-    }
-
-    @SuppressWarnings("unchecked")
-    public byte[] getSerialized() {
-      return serializedBinary =
-          serializedBinary == null
-              ? key.toBytes(parsed) : serializedBinary;
-    }
-
-    @Override
-    public String toString() {
-      if (!isBinary) {
-        return new String(getSerialized(), US_ASCII);
-      } else {
-        // Assume that the toString of an Object is better than a binary encoding.
-        if (parsed != null) {
-          return "" + parsed;
-        } else {
-          return Arrays.toString(serializedBinary);
-        }
-      }
-    }
-  }
-
-  private static class ValueIterable<T> implements Iterable<T> {
-    private final Key<T> key;
-    private final Iterable<MetadataEntry> entries;
-
-    public ValueIterable(Key<T> key, Iterable<MetadataEntry> entries) {
-      this.key = key;
-      this.entries = entries;
-    }
-
-    @Override
-    public Iterator<T> iterator() {
-      final Iterator<MetadataEntry> iterator = entries.iterator();
-      class ValueIterator implements Iterator<T> {
-        @Override public boolean hasNext() {
-          return iterator.hasNext();
-        }
-
-        @Override public T next() {
-          return iterator.next().getParsed(key);
-        }
-
-        @Override public void remove() {
-          // Not implemented to not need to conditionally update {@link #storeCount}.
-          throw new UnsupportedOperationException();
-        }
-      }
-
-      return new ValueIterator();
-    }
-  }
 }
```

Comment:
Yeah, and the logic is harder to follow now.  These iterators and entries create a tremendous amount of garbage, so getting rid of them was a goal.